File: backend/app/schemas/messaging.py
import datetime
import uuid

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.message import MessageKind
from app.schemas.auth import UserOut
from app.schemas.memes import MemeOut

MAX_MESSAGE_LENGTH = 2000


class ConversationCreate(BaseModel):
    user_id: uuid.UUID


class MessageCreate(BaseModel):
    kind: MessageKind = MessageKind.text
    body: str | None = Field(default=None, max_length=MAX_MESSAGE_LENGTH)
    meme_id: uuid.UUID | None = None

    @model_validator(mode="after")
    def _check_payload_matches_kind(self) -> "MessageCreate":
        if self.kind is MessageKind.text:
            if self.meme_id is not None:
                raise ValueError("A text message cannot carry a meme_id")
            body = (self.body or "").strip()
            if not body:
                raise ValueError("A text message needs a non-empty body")
            self.body = body
        else:
            if self.meme_id is None:
                raise ValueError("A meme message needs a meme_id")
            # A meme message is the meme; an accompanying caption would be a second
            # content channel to render, moderate and truncate for no product gain.
            if self.body is not None:
                raise ValueError("A meme message cannot carry a body")
        return self


class MessageOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    conversation_id: uuid.UUID
    sender: UserOut
    kind: MessageKind
    body: str | None
    # Null on a text message, and also on a meme message whose meme was deleted after
    # the fact — the client renders an unavailable-attachment placeholder for the latter.
    meme: MemeOut | None
    read_at: datetime.datetime | None
    created_at: datetime.datetime


class ConversationOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    other_user: UserOut
    last_message: MessageOut | None
    unread_count: int
    last_message_at: datetime.datetime | None


class MessagePage(BaseModel):
    items: list[MessageOut]
    next_cursor: str | None


class ConversationReadOut(BaseModel):
    conversation_id: uuid.UUID
    read_count: int
    read_at: datetime.datetime | None

File: frontend/src/services/messaging.ts
import { api } from '@/services/api';
import type { AuthUserResponse } from '@/services/auth';
import type { MemeResponse } from '@/services/memes';

export type MessageKind = 'text' | 'meme';

export interface MessageResponse {
  id: string;
  conversation_id: string;
  sender: AuthUserResponse;
  kind: MessageKind;
  body: string | null;
  /** Null for text messages, and for a meme message whose meme was deleted since. */
  meme: MemeResponse | null;
  read_at: string | null;
  created_at: string;
}

export interface ConversationResponse {
  id: string;
  other_user: AuthUserResponse;
  last_message: MessageResponse | null;
  unread_count: number;
  last_message_at: string | null;
}

export interface MessagePageResponse {
  items: MessageResponse[];
  next_cursor: string | null;
}

export interface ConversationReadResponse {
  conversation_id: string;
  read_count: number;
  read_at: string | null;
}

export type SendMessagePayload =
  | { kind: 'text'; body: string }
  | { kind: 'meme'; meme_id: string };

export function listConversationsRequest() {
  return api.get<ConversationResponse[]>('/messaging/conversations');
}

export function openConversationRequest(userId: string) {
  return api.post<ConversationResponse>('/messaging/conversations', { user_id: userId });
}

export function listMessagesRequest(conversationId: string, cursor: string | null, limit = 30) {
  return api.get<MessagePageResponse>(`/messaging/conversations/${conversationId}/messages`, {
    limit,
    ...(cursor ? { cursor } : {}),
  });
}

export function sendMessageRequest(conversationId: string, payload: SendMessagePayload) {
  return api.post<MessageResponse>(
    `/messaging/conversations/${conversationId}/messages`,
    payload
  );
}

export function markConversationReadRequest(conversationId: string) {
  return api.post<ConversationReadResponse>(`/messaging/conversations/${conversationId}/read`);
}
